src/shims/foreign_items/posix.rs

```diff
@@ -23,40 +24,51 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         match link_name {
             // Environment related shims
             "getenv" => {
+                check_arg_count(args, 1)?;
                 let result = this.getenv(args[0])?;
                 this.write_scalar(result, dest)?;
             }
             "unsetenv" => {
+                check_arg_count(args, 1)?;
                 let result = this.unsetenv(args[0])?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "setenv" => {
+                check_arg_count(args, 3)?;
                 let result = this.setenv(args[0], args[1])?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "getcwd" => {
+                check_arg_count(args, 2)?;
                 let result = this.getcwd(args[0], args[1])?;
                 this.write_scalar(result, dest)?;
             }
             "chdir" => {
+                check_arg_count(args, 1)?;
                 let result = this.chdir(args[0])?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
 
             // File related shims
             "open" | "open64" => {
+                check_arg_count(args, 3)?;
                 let result = this.open(args[0], args[1])?;
                 this.write_scalar(Scalar::from_i32(result), dest)?;
             }
             "fcntl" => {
+                if !(args.len() == 2 || args.len() == 3) {
+                    throw_ub_format!("incorrect number of arguments, got {}, needed 2 or 3", args.len());
+                }
```

Comment:
This seems oddly permissive... it seems better to pass `&args[2..]` to `this.fcntl` and then in there do the proper validation for each possible command.